kotlin/internal/jvm/compile.bzl

```diff
@@ -359,6 +380,198 @@ def _run_merge_jdeps_action(ctx, toolchains, jdeps, outputs, deps):
         toolchain = _TOOLCHAIN_TYPE,
     )
 
+def _run_ksp2_standalone_action(
+        ctx,
+        rule_kind,
+        toolchains,
+        srcs,
+        compile_deps,
+        annotation_processors,
+        transitive_runtime_jars):
+    """Runs KSP2 in standalone mode (not as compiler plugin).
+
+    This is simpler than compiler plugin mode:
+    1. Invoke KSP2 as a separate Java program
+    2. KSP2 generates sources to disk
+    3. Package generated sources into srcjars
+    4. Main compilation will include these sources
+
+    Args:
+        ctx: Rule context
+        rule_kind: Kind of rule (e.g., "kt_jvm_library")
+        toolchains: Kotlin toolchains
+        srcs: Source files (kt, java)
+        compile_deps: Compilation dependencies
+        annotation_processors: KSP annotation processors
+        transitive_runtime_jars: Runtime JARs
+
+    Returns:
+        struct with:
+            kotlin_srcjar: Generated Kotlin sources
+            java_srcjar: Generated Java sources
+    """
+
+    # Declare output directories
+    kotlin_out_dir = ctx.actions.declare_directory(
```

Comment:
What is the advantage of all these directories as outputs?

I ask because `delcare_directory` is not only inefficient, but is only useful in very specific circumstances -- when you do not know the output of an action, but need to get specific artifacts out of it.